### ISSUE TITLE
Database exceeds 16k leaf pages

### ISSUE THREAD

**Issue by janstarke** [libyal/libesedb#71]
I'm currently handling a PR for a mitigation of said problem: janstarke/ntdsextract2#16

There seems to be a situation in which the first call to `libesedb_record_get_number_of_values()` fails, while a subsequent call to the same function succeeds. Can you imagine a situation in which this might happen?

Unfortunately, it is not possible to provide the data which causes this behaviour :cry:

**Comment by joachimmetz** [comment]
> Unfortunately, it is not possible to provide the data which causes this behaviour

can you provide debug information instead?

**Comment by janstarke** [comment]
All I have is the error message:

```
Error: c-libesedb: libesedb_table_get_number_of_records: unable to retrieve number of leaf values from table page tree.
libesedb_page_tree_get_number_of_leaf_values: invalid number of leaf pages value exceeds maximum.
libesedb_table_get_number_of_records: unable to retrieve number of leaf values from table page tree.
```

@Darkrael, you have the test data. Could you please provide some more information to @joachimmetz to help him finding the issue?

Regards, Jan

**Comment by joachimmetz** [comment]
Also what is `c-libesedb` ?

**Comment by janstarke** [comment]
> Also what is `c-libesedb` ? That seems to be generating the warning.

this is simply a static string in `rust-libesedb`:

https://github.com/sunsetkookaburra/rust-libesedb/blob/6b88c3a440de2de7f1e660c8edc1a6cdcc16e7ec/src/error.rs#L44-L58

`rust-libesedb` itself is a rust binding for libesedb

**Comment by Darkrael** [comment]
> > Unfortunately, it is not possible to provide the data which causes this behaviour
> 
> can you provide debug information instead?

Hey, are you talking about these building flags: https://github.com/libyal/libesedb/wiki/Building#verbose-and-debug-output? If so i can try them. I will get back to you in the coming week when i find the time

**Comment by Darkrael** [comment]
@joachimmetz I've sent you an e-mail to the mail in your profile with some debug files. I hope it helps

**Comment by joachimmetz** [comment]
Finally got around to this. 

Based on:
```
libesedb_page_tree_get_number_of_leaf_values: invalid number of leaf pages value exceeds maximum.
```

https://github.com/libyal/libesedb/blob/768a0474f1eb4c594d18886b7361bcb92b139c07/libesedb/libesedb_page_tree.c#L1739

So it looks like the database has more than 16k leaf pages and based on the debug output looks like this limit is being hit.

```
grep 'current page number' filtered_datatable.txt | wc -l
16387
```

Tldr database is too big for current hard limit in libesedb. Would need to take a closer look at updating LIBESEDB_MAXIMUM_NUMBER_OF_LEAF_PAGES and memory consumption.

@Darkrael if you still have the database you could try increasing LIBESEDB_MAXIMUM_NUMBER_OF_LEAF_PAGES

**Comment by Darkrael** [comment]
Hey thanks for looking further into this. Unfortunatly i do not have the database anymore. If i encounter this error again in another database, i will try this.

**Comment by joachimmetz** [comment]
Thanks, I'll close this issue then. Given that extending the number of leaf pages would be part of handling large(r) databases https://github.com/libyal/libesedb/issues/2
